ingress/controllers/gce/storage/pools.go

```diff
@@ -51,3 +56,78 @@ func NewInMemoryPool() *InMemoryPool {
 	return &InMemoryPool{
 		cache.NewThreadSafeStore(cache.Indexers{}, cache.Indices{})}
 }
+
+type keyFunc func(interface{}) (string, error)
+
+type cloudLister interface {
+	List() ([]interface{}, error)
+}
+
+// CloudListingPool wraps InMemoryPool but relists from the cloud periodically.
+type CloudListingPool struct {
+	// A lock to protect against concurrent mutation of the pool
+	lock sync.Mutex
+	// The pool that is re-populated via re-list from cloud, and written to
+	// from controller
+	*InMemoryPool
+	// An interface that lists objects from the cloud.
+	lister cloudLister
+	// A function capable of producing a key for a given object.
+	// This key must match the key used to store the same object in the user of
+	// this cache.
+	keyGetter keyFunc
+}
+
+// ReplinishPool lists through the cloudLister and inserts into the pool.
+func (c *CloudListingPool) ReplinishPool() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	glog.V(4).Infof("Replinishing pool")
+	items, err := c.lister.List()
+	if err != nil {
+		glog.Warningf("Failed to list: %v", err)
+		return
+	}
+	for i := range items {
+		key, err := c.keyGetter(items[i])
+		if err != nil {
+			glog.V(4).Infof("CloudListingPool: %v", err)
+			continue
+		}
+		c.InMemoryPool.Add(key, items[i])
```

Comment:
Delete happens in sync. The point of this is so we don't leak, deleting from this pool would defeat the purpose, eg:
1. Create ingress foo
2. Create fw rule for foo
3. Container stops
4. While container is down, someone deletes ingress foo
5. Container starts
6. Container syncs existing ingress with cloud, ignores rule foo because there is no corresponding ingress

This ReplishPool will add the foo fw rule into the pool so when we sync we take a diff and note that there's an extra foo fw rule without a corresponding ingress that needs deletion. 

Everything in the pool is reconfirmed periodically, so eg: the controller adds fw rule foo to the pool, a user goes to cloud console and deletes foo (this would be a legit case where we need to delete here), the periodic sync of Ingress will reconfirm that foo doesn't exist, create it, and update the pool.